src/llvm-ptls.cpp

```diff
@@ -0,0 +1,185 @@
+// This file is a part of Julia. License is MIT: http://julialang.org/license
+
+#define DEBUG_TYPE "lower_ptls"
+#undef DEBUG
+
+// This file defines two entry points:
+//     global function annotateSimdLoop: mark a loop as a SIMD loop.
+//     createLowerSimdLoopPass: construct LLVM for lowering a marked loop later.
+
+#include "llvm-version.h"
+#include "support/dtypes.h"
+
+#include <llvm/Pass.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/LLVMContext.h>
+
+#include "julia.h"
+#include "julia_internal.h"
+
+#if defined(LLVM37) && defined(JULIA_ENABLE_THREADING)
+#  include <llvm/IR/InlineAsm.h>
+#endif
+
+using namespace llvm;
+
+namespace {
+
+struct LowerPTLS: public ModulePass {
+    static char ID;
+    LowerPTLS(bool _imaging_mode=false, MDNode *_tbaa_const=nullptr)
+        : ModulePass(ID),
+          imaging_mode(_imaging_mode),
+          tbaa_const(_tbaa_const)
+    {}
+
+private:
+    bool imaging_mode;
+    MDNode *tbaa_const; // One `LLVMContext` only
+    bool runOnModule(Module &M) override;
+    void runOnFunction(LLVMContext &ctx, Module &M, Function *F,
+                       Function *ptls_getter, Type *T_ppjlvalue);
+};
+
+static void ensure_global(const char *name, Type *t, Module &M)
+{
+    if (M.getNamedValue(name))
+        return;
+    GlobalVariable *proto = new GlobalVariable(M, t, false,
+                                               GlobalVariable::ExternalLinkage,
+                                               NULL, name);
+#ifdef LLVM35
+    // DLLImport only needs to be set for the shadow module
+    // it just gets annoying in the JIT
+    proto->setDLLStorageClass(GlobalValue::DefaultStorageClass);
+#endif
```

Comment:
no, it's permanently gone. besides, we don't need comdat on variables or dllimports.

but that does bring up the point that this declaration will be wrong for a no-threads build on windows, since at the end of the pass, it'll be missing the DLLImport attribute.